core/modules/url/store/actions.ts

```diff
@@ -0,0 +1,77 @@
+import { UrlState } from '../types/UrlState'
+import { ActionTree } from 'vuex';
+import * as types from './mutation-types'
+// you can use this storage if you want to enable offline capabilities
+import { cacheStorage } from '../'
+import { parseURLQuery } from '@vue-storefront/core/helpers'
+import SearchQuery from '@vue-storefront/core/lib/search/searchQuery'
+
+// it's a good practice for all actions to return Promises with effect of their execution
+export const actions: ActionTree<UrlState, any> = {
+  // if you want to use cache in your module you can load cached data like this
+  registerMapping ({ commit }, { url, routeData }) {
+    return new Promise ((resolve, reject) => {
+      commit(types.REGISTER_MAPPING, { url, routeData })
+      cacheStorage.setItem(url, routeData).then(result => {
+        resolve(routeData)
+      }).catch(() => reject())
+    })
+  },
+  mapUrl ({ state, dispatch }, { url, query }) {
+    const parsedQuery = typeof query === 'string' ? parseURLQuery(query) : query
+    if (url && url[0] === '/') url = url.slice(1)
+    const queryPos = url.indexOf('?')
+    if (queryPos > 0) url = url.slice(0, queryPos)
+
+    return new Promise ((resolve, reject) => {
+      if (state.dispatcherMap.hasOwnProperty(url)) {
+        return resolve (state.dispatcherMap[url])
+      }
+      cacheStorage.getItem(url).then(routeData => {
+        if (routeData !== null) {
+          return resolve(routeData)
+        } else {
+          dispatch('mappingFallback', { url, params: parsedQuery }).then(resolve).catch(reject)
+        }
+      }).catch(reject)
+    })
+  },
+  /**
+   * Router mapping fallback - get the proper URL from API
+   * This method could be overriden in custom module to provide custom URL mapping logic
+   */
+  mappingFallback ({ commit, dispatch }, { url, params }) {
```

Comment:
Here is the fallback if route hasn’t been cached we’re mapping url to route using server call - by default request for category/product are separated but it can be optimized with just one call to /api/mapUrl - or something like this. Here is the source for where the urls come from @filrak 

Please analyze this solution carefully and propose other approach as really: there are not many different options for having this feature. At last from my r&d. I’m totally frustrated with this feature to be honest. Would love if somebody else will do this better way be cause I’m also not very happy with the overall complication and of the duplicates logic of client-entry.ts vs UrlClientDispatcher.ts.

I’ve had some partial successes with other approaches but always there was some deal breaker. The biggest problem was with not executing the full Vue-router flow in case when route has not changed. Imagine the situation when user request is handled by the by “urldisparcher” and user then clicks another time product link. Vue is just executing the watch on $route inside current component. In case if you are on product component clicking on category link you have no way to lazy load category component and replace the current one ....